### PR TITLE
Line-of-sight structure-enhanced ray-tracing halos rendering for external convergence and shear

### PR DESCRIPTION
**Pull request by trivialTZ**
### **Line-of-Sight Structure-Enhanced Ray-Tracing Halos Rendering for External Convergence and Shear**

**Description:**

This pull request introduces an enhancement to the ray-tracing halos rendering mechanism. It incorporates both large-scale structures using GLASS and small-scale structures via the halo mass function.

##### Key Features:

1. **Line-of-Sight Structure Rendering**: This implementation utilizes the SkyPy halo module for the line-of-sight structure. It's worth noting that the currently used SkyPy halo module has not been merged into the main SkyPy branch, hence we're temporarily making use of `Skypy_halos_duplicate`.
2. **Ray-Tracing Computations**: For a given sky area, this feature generates halos within the corresponding light cone. It then employs ray-tracing techniques to compute the distribution of convergence ( $\kappa_{ext}$ ) and shear ( $\gamma_{tot}$ ).
3. **Negative Mass Sheet Option**: Users have the flexibility to generate a negative mass sheet which can be used for kappa correction.
4. **Demonstrative Notebook**: A notebook has been added to provide a hands-on demonstration of this feature.

**Next Steps:**

1. **Testing and Documentation**:
   - **Test Functions**: Enhance and finalize the test functions to ensure the feature's robustness.
   - **Notebook**: Refine the notebook to provide clearer demonstrations and examples.
   - **Documentation**: Improve the documentation to ensure comprehensive understanding and ease of use for the users.

2. **Incorporate $\kappa_{ext}$ Correction**:
   - We aim to introduce a correction factor for $\kappa_{ext}$:
  
$$ 1 - \kappa_{ext} = \frac{(1-\kappa_d)(1-\kappa_s)}{1-\kappa_{ds}} $$
   
   
   This will ensure a more accurate representation and calculation in our rendering mechanism.

**Acknowledgements**:

- Guidance and assistance from: @sibirrer  .
- Feedback and suggestions from: @ajshajib .
- Additional help from: @nkhadka21 .

We kindly request a review of this feature, and are open to further feedback and suggestions.

### CODE REVIEW

File: sim_pipeline/Halos/halos_lens.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        -----
+        This function assumes the method get_convergence_shear() is implemented, and it returns a 4-tuple:
+        (kappa, gamma1, gamma2, gamma). If gamma parameter is False, gamma1 and gamma2 are stored,
+        otherwise gamma is stored. All returned values from get_convergence_shear() are assumed to be floats.
+        """
+
+        kappa_gamma_distribution = np.empty((self.samples_number, 2 if gamma_tot else 3))
+
+        loop = range(self.samples_number)
+        if self.samples_number > 999:
```
Comment (sibirrer) [inline sim_pipeline/Halos/halos_lens.py:380]: is there a specific reason for > 999?

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thanks a lot @trivialTZ for this PR!
there are quite some untested code blocks. Parts might be that the Skypy code is not used and could potentially be removed from the PR, and the other is lacking of test functions.
We need great test coverages for all PRs.
Maybe I haven't seen it. I don't see the total kappa calculation in this PR as described in the description (it's ok if it's not implemented, just state then that this is TODO and we should not review this part yet).

**Comment by nkhadka21** [comment]
Hi @ajshajib, I noticed that pre-commit is complaining for the lines with characters greater than 88. Should pre-commit bot fix this automatically? It seems like it is not fixing them.

**Comment by trivialTZ** [comment]
> 2\. Incorporate κext Correction

Yeah, as stated in the description. Incorporating $\kappa_{ext}$ correction is what need to be focused in next step.

**Pull request merged**